trace-dispatcher/doc/trace-dispatcher.md

```diff
@@ -0,0 +1,798 @@
+* [ ] [Decide on more direct interface to EKG metrics](#Decide-on-more-direct-interface-to-ekg-metrics)
+* [x] [Decide on dispatcher detail level control](#Decide-on-dispatcher-detail-level-control)
+* [x] [Decide namespace-aware configuration](#Decide-namespace-aware-configuration)
+* [ ] [Discuss impact of missing documentation entries](#Discuss-impact-of-missing-documentation-entries)
+* [x] [Decide missing configuration](#Decide-missing-configuration)
+* [x] [Decide complete configuration](#Decide-complete-configuration)
+* [ ] [Discuss possibility of pure, thread-safe aggregation](#Discuss-possibility-of-pure-thread-safe-aggregation)
+* [ ] [Decide trace-outs types](#Decide-trace-outs-types)
+* [ ] Agree on editing.
+* [ ] Final proofreading.
+* [ ] Feedback.
+
+# Introduction
+
+## Motivation
+
+`trace-dispatcher` is an attempt to enable design and implementation of simple, efficient __tracing systems__, one that has a reduced footprint in the executed program, has a more pleasant API, and provides self-documenting features.
+
+## Design decisions
+
+Key design decisions were:
+
+1. Retaining the separation of concerns in the frontend side, as provided by the `contra-tracer` library.  The client code should not need to concern itself with any details beyond passing the traces down to the system.
+2. Rely on __trace combinators__ primarily, as opposed to opting for a typeclass heavy API.
+3. Separation of data plane and control plane:  high-frequency events incur minimal processing on the data-plane, whereas complicated configuration logic only happens on the control plane, and that is proportional to infrequent reconfiguration events.
+4. A tougher stance on separation of concerns in the backend side:  we choose to move expensive trace processing to an external process.
+5. A measure of backward compatibility with the previous logging system.
+6. Retaining the global namespace for all traces.
+
+## Overview and terminology
+
+The emitted __program traces__ (streams of __messages__ of arbitrary data types, where each data type defines a number of different __messages__) are collected across all program components, and undergo __trace interpretation__ by the __dispatcher__ into __metrics__ and __messages__, which are afterwards externalised.
+
+Therefore, we can conceptually decompose the __tracing system__ into three components:
+
+* __frontend__, the entry point for __program trace__ collection, which is just a single function `traceWith`;  Program locations that invoke this frontend (thereby injecting messages into the tracing system) is called __trace-ins__.
+* __dispatcher__, is a structured, namespaced set of contravariantly-composed transformations, triggered by the entry point.  Its role is specifically __trace interpretation__;
+* __backend__, externalises results of the interpretation ( __metrics__ and __messages__) outside the system, through __trace-outs__.
```

Comment:
By "system" do you mean "process"? If yes, perhaps stick with "process"  since you've used that wording already.

Reply:
No, clarifying that to be tracing system, in the sense of a library, infrastructure component.